content/guides/typescript-for-solid.mdx

```diff
@@ -0,0 +1,552 @@
+
+```typescript
+let divRef!: HTMLDivElement;
+// This will be permitted but will throw an error at runtime
+divRef.focus();
+onMount(() => {
+  divRef.focus(); 
+});
+return <div ref={divRef}>...</div>;
+```
+
+## Control flow narrowing
+
+When using control flow components in Solid, such as [`<Show>`](/reference/show), it's common to conditionally display content based on the presence of certain data.
+This, however, can introduce some type-checking challenges, especially when attempting to manipulate the daata within these components.
+
+Consider a situation where you want to show a name only when it's defined using a Solid signal:
+
+```typescript
+const [name, setName] = createSignal<string>();
+return <Show when={name()}>Hello {name().replace(/\s+/g, "\xa0")}!</Show>;
+```
+
+TypeScript may not like this.
+While you can understand that the second call to `name()` happens only if the first is truthy, TypeScript cannot. 
+It may flag an error indicating that `name()` could potentially be `undefined` when `.replace()` is called.
+
+There are a few ways to tackle this:
+
+1. **Non-Null Assertion**
+Use TypeScript's [non-null assertion operator (!).](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) 
+This operator tells TypeScript you're confident the value isn't `null` or `undefined`:
+
+```typescript
+return <Show when={name()}>Hello {name()!.replace(/\s+/g, "\xa0")}!</Show>;
+```
+While this may be a quick option, incorrectly asserting can lead to runtime errors if the value does turn out to be `null` or `undefined`.
+
+2. **Callback form of Show**
+
+Solid offers a callback form for the [`<Show>`](/reference/show) component, where the truthy value of the `when` prop gets passed in:
+
+```typescript
+return (
+  <Show when={name()}>{(n) => <>Hello {n.replace(/\s+/g, "\xa0")}!</>}</Show>
+);
+```
+
+This method leverages the type information embedded in the [`<Show>`](/reference/show) component.
+TypeScript can infer that since `n` comes from a truthy `name()`, it can't be `undefined` or other falsy values.
+However, accessing this null asserted form incorrectly may thrown an error if the condition is not satisfied.
+
+## Advanced JSX attributes and directives
+
+### Custom event handlers
+
+For custom events in Solid, the [`on:___`]() and [`oncapture:___`]() attributes can be utilized. 
+To type these events, you'll need to extend Solid's JSX namespace.
+
+```typescript
+class NameEvent extends CustomEvent {
+  type: "Name";
+  detail: { name: string };
+
+  constructor(name: string) {
+    super("Name", { detail: { name } });
+  }
+}
+
+declare module "solid-js" {
+  namespace JSX {
+    interface CustomEvents {
+      Name: NameEvent;  // Matches `on:Name`
+    }
+    interface CustomCaptureEvents {
+      Name: NameEvent;  // Matches `oncapture:Name`
+    }
+  }
+}
```

Comment:
You might also include a way to extend custom events to include native html events too. This might be helpful if you want to attach a normal event-listener, instead of relying on event delegation.
By default the native events are not included, so doing `<div on:mousemove={e => {}} />` will result in a typescript error.
They could be easily added by doing:
```ts
declare module "solid-js" {
  namespace JSX {
    interface CustomEvents extends HTMLElementEventMap {}
    interface CustomCaptureEvents extends HTMLElementEventMap {}
  }
}
```
Or to pick only the events we need:
```ts
declare module "solid-js" {
  namespace JSX {
    interface CustomEvents extends Pick<HTMLElementEventMap, "mousemove" | "pointermove"> {}
  }
}
```

Reply:
Ended up adding this in an aside - thank you!